pkg/parse/events/builder.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package events
+
+import (
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/utils/clock"
+)
+
+// PublishingGroupBuilder oversees construction of event publishers.
+//
+// For now, the publishers are driven by clock-based delay and backoff timers.
+type PublishingGroupBuilder struct {
+	// Clock is used for time tracking, namely to simplify testing by allowing
+	// a fake clock, instead of a RealClock.
+	Clock clock.Clock
+	// SyncPeriod is the period of time between checking the filesystem
+	// for publisher updates to sync.
+	SyncPeriod time.Duration
+	// SyncWithReimportPeriod is the period of time between forced re-sync from
+	// publisher (even without a new commit).
+	SyncWithReimportPeriod time.Duration
+	// StatusUpdatePeriod is how long the Parser waits between updates of the
+	// sync status, to account for management conflict errors from the Remediator.
+	StatusUpdatePeriod time.Duration
+	// NamespaceControllerPeriod is how long to wait between checks to see if
+	// the namespace-controller wants to trigger a resync.
+	// TODO: Use a channel, instead of a timer checking a locked variable.
+	NamespaceControllerPeriod time.Duration
+	// RetryBackoff is how long the Parser waits between retries, after an error.
+	RetryBackoff wait.Backoff
+}
+
+// Build a list of Publishers based on the PublishingGroupBuilder config.
+func (t *PublishingGroupBuilder) Build() []Publisher {
+	var publishers []Publisher
+	if t.SyncPeriod > 0 {
+		publishers = append(publishers, NewResetOnRunAttemptPublisher(SyncEventType, t.Clock, t.SyncPeriod))
+	}
+	if t.SyncWithReimportPeriod > 0 {
+		publishers = append(publishers, NewTimeDelayPublisher(SyncWithReimportEventType, t.Clock, t.SyncWithReimportPeriod))
+	}
+	if t.NamespaceControllerPeriod > 0 {
+		publishers = append(publishers, NewTimeDelayPublisher(NamespaceResyncEventType, t.Clock, t.NamespaceControllerPeriod))
+	}
+	if t.RetryBackoff.Duration > 0 {
+		publishers = append(publishers, NewRetrySyncPublisher(t.Clock, t.RetryBackoff))
+	}
+	if t.StatusUpdatePeriod > 0 {
+		publishers = append(publishers, NewResetOnRunAttemptPublisher(StatusEventType, t.Clock, t.StatusUpdatePeriod))
```

Comment:
Does StatusEventType need a ResetOnRunAttemptPublisher? `RunAttempt` will always be false for StatusEvent, so it will never get delayed.

Reply:
ACK. It also allows the retry backoff to be reset.

Reply:
retry backoff is reset by a different flag: Result.ResetRetryBackoff

The RunAttempted flag just resets the status update and sync timers.

Should we make the RunAttempted flag ALSO reset the retry backoff?

Reply:
Nevermind. You just meant that the HandleResult allowed for ResetRetryBackoff to be sent from other events.